src/call/buffers.rs

```diff
@@ -0,0 +1,570 @@
+    grpc_slice_refcount_vtable,
+};
+
+#[cfg(feature = "prost-codec")]
+use bytes::{Buf, BufMut};
+
+// A wrapper for `grpc_slice`.
+struct GrpcSlice(grpc_slice);
+
+impl GrpcSlice {
+    fn len(&self) -> usize {
+        unsafe { grpc_sys::grpcwrap_slice_length(&self.0) }
+    }
+
+    fn range_from(&self, offset: usize) -> &[u8] {
+        unsafe {
+            let mut len = 0;
+            let ptr = grpc_sys::grpcwrap_slice_raw_offset(&self.0, offset, &mut len);
+            slice::from_raw_parts(ptr as _, len)
+        }
+    }
+}
+
+impl Default for GrpcSlice {
+    fn default() -> Self {
+        GrpcSlice(unsafe { grpc_sys::grpc_empty_slice() })
+    }
+}
+
+impl Drop for GrpcSlice {
+    fn drop(&mut self) {
+        unsafe {
+            grpc_sys::grpcwrap_slice_unref(&self.0);
+        }
+    }
+}
+
+/// A wrapper for `grpc_byte_buffer_reader`.
+///
+/// Manages the `byte_buffer` behind the reader as well as the reader itself. That
+/// means we expect a 1:1 relationship between the `byte_buffer` and reader, so
+/// this is not a general purpose wrapper.
+struct GrpcByteBufferReader(grpc_byte_buffer_reader);
+
+impl GrpcByteBufferReader {
+    // Takes ownership of `buf` and will destroy it (at some point). The caller
+    // must not keep a reference.
+    //
+    // Safety: `buf` must be valid and non-null. The caller must not keep a reference
+    // to it.
+    unsafe fn new(buf: *mut grpc_byte_buffer) -> GrpcByteBufferReader {
+        let mut reader = MaybeUninit::uninit();
+        let init_result = grpc_sys::grpc_byte_buffer_reader_init(reader.as_mut_ptr(), buf);
+        assert_eq!(init_result, 1);
+        GrpcByteBufferReader(reader.assume_init())
+    }
+
+    fn len(&self) -> usize {
+        unsafe { grpc_sys::grpc_byte_buffer_length(self.0.buffer_out) }
+    }
+
+    fn next_slice(&mut self) -> GrpcSlice {
+        unsafe {
+            let mut slice = GrpcSlice::default();
+            let code = grpc_sys::grpc_byte_buffer_reader_next(&mut self.0, &mut slice.0);
+            debug_assert_ne!(code, 0);
+            slice
+        }
+    }
+}
+
+impl Drop for GrpcByteBufferReader {
+    fn drop(&mut self) {
+        unsafe {
+            let buf = self.0.buffer_in;
+            grpc_sys::grpc_byte_buffer_reader_destroy(&mut self.0);
+            grpc_sys::grpc_byte_buffer_destroy(buf);
+        }
+    }
+}
+
+/// Constructor function for a `grpc_slice` which wraps a Rust `Vec`.
+///
+/// Safety: see `VecSliceRefCount::new`, the same applies here.
+unsafe fn vec_slice(v: Vec<u8>) -> Box<grpc_slice> {
+    let mut data: MaybeUninit<grpc_sys::grpc_slice_grpc_slice_data> = MaybeUninit::zeroed();
+    *(*data.as_mut_ptr()).refcounted.as_mut() = grpc_sys::grpc_slice_grpc_slice_data_grpc_slice_refcounted {
+        bytes: v.as_ptr() as *const _ as *mut _,
+        length: v.len(),
+    };
+    let mut refcount = VecSliceRefCount::new(v);
+    let mut result = Box::new(grpc_slice {
```

Comment:
Why box the slice?

Reply:
I think we have to - we return it from the function so we can't stack allocate. I experimented with arena allocating it, but it ended up being slower.

Reply:
grpc_slice is OK to move around, so I don't see why can't it be stack allocated.

Reply:
Because the caller of this function ends up moving it into `grpc_raw_byte_buffer_create`, which takes a `*mut grpc_slice` and will outlive the function it is used in, so we have to heap allocate at some point. We could return the `vec_slice` by value here, but it ends up on the heap

Reply:
`grpc_raw_byte_buffer_create` won't move grpc_slice. It refs/clones the slices. Ref: https://grpc.github.io/grpc/core/impl_2codegen_2byte__buffer_8h.html#a5ab72d6615ee89c975c46ad4e50c1387

The detail implement is here: 

https://github.com/grpc/grpc/blob/cc43fd64ab97e45fcdd38759dd8d9eb273823b9e/src/core/lib/surface/byte_buffer.cc#L34-L47

Reply:
you are right, I think it can be simplified...